embassy-stm32/src/ucpd.rs

```diff
@@ -0,0 +1,218 @@
+//! USB Type-C/USB Power Delivery Interface (UCPD)
+
+// Implementation Notes
+//
+// As of Feb. 2024 the UCPD peripheral is availalbe on: G0, G4, H5, L5, U5
+//
+// Cube HAL LL Driver (g0):
+// https://github.com/STMicroelectronics/stm32g0xx_hal_driver/blob/v1.4.6/Inc/stm32g0xx_ll_ucpd.h
+// https://github.com/STMicroelectronics/stm32g0xx_hal_driver/blob/v1.4.6/Src/stm32g0xx_ll_ucpd.c
+// Except for a the `LL_UCPD_RxAnalogFilterEnable/Disable()` functions the Cube HAL implementation of
+// all families is the same.
+//
+// Dead battery pull-down resistors functionality is enabled by default on startup and must
+// be disabled by setting a bit in PWR/SYSCFG registers. The exact name and location for that
+// bit is different for each familily.
+
+use core::future::poll_fn;
+use core::marker::PhantomData;
+use core::task::Poll;
+
+use crate::rcc::RccPeripheral;
+use crate::{interrupt, pac};
+use embassy_hal_internal::{into_ref, Peripheral, PeripheralRef};
+use embassy_sync::waitqueue::AtomicWaker;
+use pac::ucpd::vals::{Anamode, Ccenable, PscUsbpdclk};
+
+pub use pac::ucpd::vals::TypecVstateCc as CcVState;
+
+/// Pull-up or Pull-down resistor state of both CC lines.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum CcPull {
+    /// Analog PHY for CC pin disabled.
+    Disabled,
+
+    /// Rd=5.1k pull-down resistor enabled when the corresponding DBCC pin is high.
+    SinkDeadBattery,
+
+    /// Rd=5.1k pull-down resistor.
+    Sink,
+
+    /// Rp=56k pull-up resistor to indicate default USB power.
+    SourceDefaultUsb,
+
+    /// Rp=22k pull-up resistor to indicate support for up to 1.5A.
+    Source1_5A,
+
+    /// Rp=10k pull-up resistor to indicate support for up to 3.0A.
+    Source3_0A,
+}
+
+/// UCPD driver.
+pub struct Ucpd<'d, T: Instance> {
+    _peri: PeripheralRef<'d, T>,
+}
+
+impl<'d, T: Instance> Ucpd<'d, T> {
+    /// Creates a new UCPD driver instance.
+    pub fn new(
+        peri: impl Peripheral<P = T> + 'd,
+        _cc1: impl Peripheral<P = impl Cc1Pin<T>> + 'd,
+        _cc2: impl Peripheral<P = impl Cc2Pin<T>> + 'd,
```

Comment:
best to get the opinion of an actual reviewer, but could it be argued that there should be 4 constructors (cc1, cc2) (cc1, cc2, dbcc1, dbcc2), (cc1), (cc1, dbcc1)  (nb. some packages don't even break out cc2. only cc1 is needed for a "direct attach" device with a plug instead of a receptacle) and then a new_inner like spi? what's the "rule" on passing in these "additonal function" (as opposed to alternate function which need setting as alternate function in gpio MODER + AFR) that don't really need anything doing to them afaik? or maybe they need setting as analog?

Reply:
In my first version I included the dbccX (dead battery input) pins as part of the constructor but decided against that later.
From intuition I would expect the driver to reconfigure the pins I pass it in the constructor as required by the application (e.g. set up pull-up and AF mode).
With CC pins thats what we do: Configure pull resistors as required.
With the DBCC I’m not sure: We actually do not change any behavior on them. The pins are already configured as such before even passing them to the constructor.

Reply:
> nb. some packages don't even break out cc2. only cc1 is needed for a "direct attach" device with a plug instead of a receptacle
Really? ST never fails to surprise! I would expect the majority of embassy applications to use a usb-c receptacly so would like to restrict the first implementation to that and keep the CC1 only plug for a follow up PR.